### DISCUSSION TITLE
[Feature Request] Integrate Local LLM for Content Summarization in Zen

### DISCUSSION THREAD

**Discussion by KodeBarista** [zen-browser/desktop#3120] [Ideas]
The ability to quickly summarize content on a webpage can be a game-changer for productivity and information consumption. Integrating a local Large Language Model (LLM) into Zen Browser would empower users with a privacy-respecting, offline feature to condense and analyze webpage content effectively.

**Comment by jeanlucthumm** [comment]
To play devil's advocate, why not use an extension for this? What's the justification for having it built into the browser?

**Reply by KodeBarista** [reply]
What extension would you recommend?

**Comment by pinage404** [comment]
It is already possible to have it

## Setup

The following steps have to be done the first time only

1. Install [Open WebUI](https://openwebui.com/)
1. Setup Open WebUI
1. Install [Ollama](https://ollama.com/)
1. Download a model (for example `llama3.2:1b`)
	* Using Open WebUI http://localhost:8080/admin/settings/ (replace the port `8080` with the configured one)
		1. Go to `Admin Panel`
		1. Click on `Settings` tab
		1. Click on `Models` tab
		1. In the `Pull a model from Ollama.com` input, fill with `llama3.2:1b`
		1. Click on the download button
	* Or using Ollama directly
	   ```sh
       ollama pull llama3.2:1b
      ```
1. In the page `about:config`, set `browser.ml.chat.hideLocalhost` to `false`
1. In the page `about:preferences#experimental`, check `AI Chatbot`
1. In the page `about:preferences#experimental`, in `Select a chatbot`, select `localhost`

### Change default model

1. Go to Open WebUI http://localhost:8080/admin/settings/ (replace the port `8080` with the configured one)
1. Go to `Admin Panel`
1. Click on `Settings` tab
1. Click on `Users` tab
1. In `Manage Models` > `Default Model`, select your favorite model

## Daily usage

1. Go to any web page (e.g. https://zen-browser.app)
1. Select some text
1. Right click
1. Hover `Ask localhost`
1. Click on `Summarize`
1. Enjoy

---

Tested with [this NixOS configuration](https://gitlab.com/pinage404/dotfiles/-/blob/746f6a58164b64b6b83f7897a2b85f6a56f29dbd/systems/service/ollama.nix) :
Zen : 1.0.1-a.19 (Firefox 132.0.1) (64 bits)
Open WebUI : v0.3.12
Ollama : 0.4.1

**Reply by KodeBarista** [reply]
Thanks for this thorough browser-llama setup explanation

**Reply by qBitnaut** [reply]
If your endpoint isn't on the same device, say you are hosting your LLM in a private cloud or another server in your homelab, you can also edit the attribute: **browser.ml.chat.provider** in **about:config** and add your own endpoint URL.
